File: medium/Add Two Numbers/addTwoNumbers.js
const ListNode = require("../../commons/LinkedList").ListNode;

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
const addTwoNumbers = function (l1, l2) {
  const result = new ListNode();

  let pointer1 = l1;
  let pointer2 = l2;
  let pointer3 = result;

  let carry = 0;

  while (pointer1.val != null && pointer2.val != null) {
    const val1 = pointer1.val;
    const val2 = pointer2.val;

    const sum = (val1 + val2 + carry) % 10;
    carry = Math.floor((val1 + val2 + carry) / 10);

    pointer3.val = sum;
    pointer3.next = new ListNode();

    pointer1 = pointer1.next;
    pointer2 = pointer2.next;
    pointer3 = pointer3.next;
  }

  while (pointer1.val != null) {
    const val1 = pointer1.val;
    const sum = (val1 + carry) % 10;
    carry = Math.floor((val1 + carry) / 10);

    pointer3.val = sum;
    pointer3.next = new ListNode();

    pointer1 = pointer1.next;
    pointer3 = pointer3.next;
  }

  while (pointer2.val != null) {
    const val2 = pointer2.val;
    const sum = (val2 + carry) % 10;
    carry = Math.floor((val2 + carry) / 10);

    pointer3.val = sum;
    pointer3.next = new ListNode();

    pointer2 = pointer2.next;
    pointer3 = pointer3.next;
  }

  if (carry !== 0) {
    pointer3.val = carry;
    pointer3.next = new ListNode();
  }

  return result;
};

const testAddTwoNumbers = function (l1, l2, output) {
  const result = addTwoNumbers(l1, l2);
  const l3 = new ListNode();
  l3.createList(output);

  if (l3.equal(result)) {
    console.log("Test Passed!");
  } else {
    console.log(
      `Test Failed!: l1 ${l1} l2 ${l2}, result ${result} output ${output}`
    );
  }
};

const l1 = new ListNode();
const l2 = new ListNode();

l1.createList([2, 4, 3]);
l2.createList([5, 6, 4]);

testAddTwoNumbers(l1, l2, [7, 0, 8]);

l1.clear();
l2.clear();

l1.createList([0]);
l2.createList([0]);

testAddTwoNumbers(l1, l2, [0]);

l1.clear();
l2.clear();

l1.createList([9, 9, 9, 9, 9, 9, 9]);
l2.createList([9, 9, 9, 9]);

testAddTwoNumbers(l1, l2, [8, 9, 9, 9, 0, 0, 0, 1]);

File: second_pass/6_linked_list/6_add_two_numbers/add_two_numbers.py
from typing import Optional
from ..util import ListNode, LinkedList

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        s1 = l1
        s2 = l2

        val = (s1.val + s2.val) % 10
        carry = (s1.val + s2.val) // 10

        new_list = ListNode(val=val)
        s3 = new_list
        s1 = s1.next
        s2 = s2.next

        while s1 and s2:
            val = (s1.val + s2.val + carry) % 10
            carry = (s1.val + s2.val + carry) // 10
            s3.next = ListNode(val)

            s3 = s3.next
            s1 = s1.next
            s2 = s2.next

        while s1:
            val = (s1.val + carry) % 10
            carry = (s1.val + carry) // 10
            s3.next = ListNode(val)

            s3 = s3.next
            s1 = s1.next

        while s2:
            val = (s2.val + carry) % 10
            carry = (s2.val + carry) // 10
            s3.next = ListNode(val)

            s3 = s3.next
            s2 = s2.next

        if carry:
            s3.next = ListNode(carry)

        return new_list

    def testAddTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode], expected: Optional[ListNode]):
        actual = self.addTwoNumbers(l1, l2)

        if not LinkedList.are_equal(actual, expected):
            print(f"\nTest Case Failed! Linked Lists are not equal")
            print(f"actual: ")
            LinkedList.print_list(actual)
            print(f"expected: ")
            LinkedList.print_list(expected)

        else:
            print("Test Case Passed!\n")


Solution().testAddTwoNumbers(
    LinkedList.create_list([2, 4, 3]),
    LinkedList.create_list([5, 6, 4]),
    LinkedList.create_list([7, 0, 8]),
)
Solution().testAddTwoNumbers(
    LinkedList.create_list([0]),
    LinkedList.create_list([0]),
    LinkedList.create_list([0]),
)
Solution().testAddTwoNumbers(
    LinkedList.create_list([9, 9, 9, 9, 9, 9, 9]),
    LinkedList.create_list([9, 9, 9, 9]),
    LinkedList.create_list([8, 9, 9, 9, 0, 0, 0, 1]),
)
